File: backend/infrastructure/sse_ticket.py
"""
SSE ticket authentication manager.

Provides short-lived, single-use tickets for SSE connections.
Needed because the EventSource API cannot send custom headers (no JWT in URL).
"""

import logging
import secrets
import time

logger = logging.getLogger(__name__)

TICKET_TTL = 60  # Tickets expire after 60 seconds
CLEANUP_INTERVAL = 300  # Clean expired tickets every 5 minutes


class TicketData:
    """Data associated with an SSE ticket."""

    __slots__ = ("user_id", "role", "room_id", "created_at")

    def __init__(self, user_id: str, role: str, room_id: int):
        self.user_id = user_id
        self.role = role
        self.room_id = room_id
        self.created_at = time.monotonic()


class SSETicketManager:
    """
    Manages short-lived, single-use tickets for SSE authentication.

    Flow:
    1. Client calls POST /rooms/{room_id}/stream/ticket with JWT auth
    2. Server issues a ticket (random token valid for 60s)
    3. Client connects to GET /rooms/{room_id}/stream?ticket=...
    4. Server validates and consumes the ticket (single-use)
    """

    def __init__(self):
        self._tickets: dict[str, TicketData] = {}
        self._last_cleanup = time.monotonic()

    def create_ticket(self, user_id: str, role: str, room_id: int) -> str:
        """
        Create a short-lived, single-use ticket for SSE auth.

        Args:
            user_id: Authenticated user ID from JWT
            role: User role from JWT (admin/guest)
            room_id: Room the ticket is valid for

        Returns:
            Ticket string (URL-safe token)
        """
        self._maybe_cleanup()
        ticket = secrets.token_urlsafe(32)
        self._tickets[ticket] = TicketData(user_id=user_id, role=role, room_id=room_id)
        logger.debug(f"SSE ticket created for user={user_id} room={room_id}")
        return ticket

    def validate_ticket(self, ticket: str, room_id: int) -> TicketData | None:
        """
        Validate and consume a ticket (single-use).

        Args:
            ticket: The ticket string to validate
            room_id: The room ID the connection is for (must match ticket)

        Returns:
            TicketData if valid, None otherwise
        """
        data = self._tickets.pop(ticket, None)
        if data is None:
            return None

        # Check expiration
        if time.monotonic() - data.created_at > TICKET_TTL:
            logger.debug("SSE ticket expired")
            return None

        # Check room_id matches
        if data.room_id != room_id:
            logger.warning(f"SSE ticket room mismatch: ticket={data.room_id} request={room_id}")
            return None

        return data

    def _maybe_cleanup(self) -> None:
        """Periodically remove expired tickets."""
        now = time.monotonic()
        if now - self._last_cleanup < CLEANUP_INTERVAL:
            return

        self._last_cleanup = now
        expired = [
            ticket
            for ticket, data in self._tickets.items()
            if now - data.created_at > TICKET_TTL
        ]
        for ticket in expired:
            del self._tickets[ticket]

        if expired:
            logger.debug(f"Cleaned up {len(expired)} expired SSE tickets")

File: backend/src/infrastructure/sse-ticket.ts
/**
 * Single-use tickets for SSE authentication. `EventSource` cannot send an
 * `X-API-Key` header, and a JWT in the query string would land a long-lived
 * credential in every access log and `Referer`; so the client trades its JWT for
 * a ticket over an authenticated POST — 60 seconds, one use, one room.
 *
 * `auth.ts` excludes `GET /rooms/{id}/stream` from the JWT middleware precisely
 * because this module authenticates it instead.
 */

import { getLogger } from './logging/logger'
import type { UserRole } from '@/auth/roles'

const logger = getLogger('SSETicket')

/** How long a ticket stays spendable. */
export const TICKET_TTL_MS = 60_000

/** How often expired tickets are swept, at most. */
export const CLEANUP_INTERVAL_MS = 300_000

export interface TicketData {
  userId: string
  role: UserRole
  roomId: number
  /** Monotonic milliseconds, from `performance.now()`. */
  createdAt: number
}

// 32 random bytes, base64url without padding.
function generateTicket(): string {
  const bytes = crypto.getRandomValues(new Uint8Array(32))
  return Buffer.from(bytes)
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '')
}

export class SSETicketManager {
  private readonly tickets = new Map<string, TicketData>()
  private lastCleanup = performance.now()

  /** @param now Injectable clock, so tests can age a ticket without sleeping. */
  createTicket(
    userId: string,
    role: UserRole,
    roomId: number,
    now: number = performance.now(),
  ): string {
    this.maybeCleanup(now)
    const ticket = generateTicket()
    this.tickets.set(ticket, { userId, role, roomId, createdAt: now })
    logger.debug(`SSE ticket created for user=${userId} room=${roomId}`)
    return ticket
  }

  // The delete happens before every check, so a ticket is consumed even when
  // expired or aimed at another room: a replay must not get a second verdict.
  validateTicket(ticket: string, roomId: number, now: number = performance.now()): TicketData | null {
    const data = this.tickets.get(ticket)
    if (data === undefined) return null
    this.tickets.delete(ticket)

    if (now - data.createdAt > TICKET_TTL_MS) {
      logger.debug('SSE ticket expired')
      return null
    }

    if (data.roomId !== roomId) {
      logger.warning(`SSE ticket room mismatch: ticket=${data.roomId} request=${roomId}`)
      return null
    }

    return data
  }

  get size(): number {
    return this.tickets.size
  }

  // Only unspent tickets accumulate (a connected client's is removed on
  // validation), so a sweep on every issue would be the wrong trade.
  private maybeCleanup(now: number): void {
    if (now - this.lastCleanup < CLEANUP_INTERVAL_MS) return
    this.lastCleanup = now

    let removed = 0
    for (const [ticket, data] of this.tickets) {
      if (now - data.createdAt > TICKET_TTL_MS) {
        this.tickets.delete(ticket)
        removed++
      }
    }
    if (removed > 0) logger.debug(`Cleaned up ${removed} expired SSE tickets`)
  }
}
